### PR TITLE
AMR

### PR DESCRIPTION
**Pull request by RuihongQiu**
### Description

Hi,

Here is an implementation of [AMR](https://ieeexplore.ieee.org/document/8618394) model as discussed in the [CausalRec PR](https://github.com/PreferredAI/cornac/pull/415).

Thank you for your time to review the code.

### Checklist:

- [ ] I have added tests.
- [x] I have updated the documentation accordingly.
- [x] I have updated `README.md` (if you are adding a new model).
- [x] I have updated `examples/README.md` (if you are adding a new example).
- [ ] I have updated `datasets/README.md` (if you are adding a new dataset).

### CODE REVIEW

File: cornac/models/amr/recom_amr.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    init_params: dictionary, optional, default: None
+        Initial parameters, e.g., init_params = {'Bi': beta_item, 'Gu': gamma_user,
+        'Gi': gamma_item, 'Tu': theta_user, 'E': emb_matrix, 'Bp': beta_prime}
+
+    seed: int, optional, default: None
+        Random seed for weight initialization.
+
+    References
+    ----------
+    * He, R., & McAuley, J. (2016). VBPR: Visual Bayesian Personalized Ranking from Implicit Feedback.
```
Comment (tqtg) [inline cornac/models/amr/recom_amr.py:79]: Could you help change to the correct reference?

File: cornac/models/amr/recom_amr.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+            name="AMR",
+            k=10,
+            k2=10,
+            n_epochs=50,
+            batch_size=100,
+            learning_rate=0.005,
+            lambda_w=0.01,
+            lambda_b=0.01,
+            lambda_e=0.0,
+            lmd=1.0,
```
Comment (tqtg) [inline cornac/models/amr/recom_amr.py:93]: I would suggest to rename `lmd` to `lambda_adv` for readability. It's also more consistent with other regularization hyper-params.

File: cornac/models/amr/recom_amr.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    lambda_w: float, optional, default: 0.01
+        The regularization hyper-parameter for latent factor weights.
+
+    lambda_b: float, optional, default: 0.01
+        The regularization hyper-parameter for biases.
+
+    lambda_e: float, optional, default: 0.0
+        The regularization hyper-parameter for embedding matrix E and beta prime vector.
+
+    lmd: float, optional, default: 1.0
```
Comment (tqtg) [inline cornac/models/amr/recom_amr.py:57]: I don't see Eq.27 in the paper, do you mean Eq.10?

**Comment by RuihongQiu** [comment]
Hi @tqtg 

Thank you for the review. And I have changed the reference of the paper as well as the naming and the codumentation of the variable ```lambda_adv```.

**Comment by tqtg** [comment]
@RuihongQiu thanks for the updates. It looks good now. I will merge this PR.
Looking forward to your future contributions!

**Comment by RuihongQiu** [comment]
Thank you for the support and the great repo. 👍

File: cornac/models/amr/recom_amr.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+                log_likelihood = torch.nn.functional.logsigmoid(Xuij).sum()
+                
+                # adversarial part
+                feat_i.retain_grad()
+                feat_j.retain_grad()
+                log_likelihood.backward(retain_graph=True)
+                feat_i_delta = feat_i.grad
+                feat_j_delta = feat_j.grad
+                
+                adv_feat_diff = feat_diff - (feat_i_delta - feat_j_delta)
```
Comment (axhiao) [inline cornac/models/amr/recom_amr.py:238]: Hi, why is it `feat_diff - (feat_i_delta - feat_j_delta)`? Based on the equation (5) in the paper, I think it should be `feat_diff + (feat_i_delta - feat_j_delta)`. Thanks!

Comment (RuihongQiu) [reply cornac/models/amr/recom_amr.py:238]: My bad in merging repos.

Should be plus.

@tqtg

Comment (tqtg) [reply cornac/models/amr/recom_amr.py:238]: @RuihongQiu could you create a PR updating the formula?

Comment (RuihongQiu) [reply cornac/models/amr/recom_amr.py:238]: I have made an update on this [PR](https://github.com/PreferredAI/cornac/pull/429).

Please see if it is suitable.

Thank you.

Comment (tqtg) [reply cornac/models/amr/recom_amr.py:238]: Merged! 
The change will be effective in the next release. For up-to-date code, please pull the source directly from GitHub. Thanks!

**Pull request merged**